### PR TITLE
FIX for #78 to workaround ->alterCanIncludeInGoogleSitemap() not working

### LINKED ISSUE #78: 'alterCanIncludeInGoogleSitemap' not working
```php
class GoogleSitemapExtension extends DataExtension {

	....

	public function canIncludeInGoogleSitemap() {
...
		$this->owner->invokeWithExtensions('alterCanIncludeInGoogleSitemap', $can);
...
		return $can;
	}

	....
}
```

The call to `alterCanIncludeInGoogleSitemap` does not have any effect as the invokeWithExtensions `$argument` argument is a call by-value and not a call-by-reference:
```php
public function invokeWithExtensions($method, $argument = null) 
```

### PR DESCRIPTION
**Pull request by patricknelson**
Fixes #78. This is due to core bug with `->invokeWithExtensions()` not allowing references.  @tractorcow is already addressing a fix in that method (see here for details: https://github.com/silverstripe/silverstripe-framework/issues/3745#issuecomment-187941032), so in the meantime we can fix this here by simply using the return value instead of depending on the reference being updated. 

This should be _reasonably_ safe (but not 100%) since anyone using this will likely NOT be returning and modifying by reference does nothing anyway, so no loss, only gain. Only potential may be if someone is erroneously returning when they should be modifying by reference but, I hope that's unlikely :smile: 

**Question** for either @wilr or @tractorcow: Since we don't really know the developers intent (if/when they use this custom `->alterCanIncludeInGoogleSitemap()` method) should we just use the **last** return value (since one was now positively returned and technically now has intent to override). This is (almost) the same as modifying by reference, but not quite (since by reference, each additional call can look at the current value and decide to change it). However, due to issues with separation of concerns, there's no guarantee that each observer of that `->alterCanIncludeInGoogleSitemap()` method will conspire to "or" vs "and" the value (since it is just a boolean).

**Another question:** Since the concept of altering is effectively useless until `silverstripe-framework` v4 rolls around (with the fix being in master) will it make sense to tweak the name to not include the word "alter"?

### CODE REVIEW

File: code/extensions/GoogleSitemapExtension.php [hunk 1]
```diff
@@ -33,7 +33,8 @@ public function canIncludeInGoogleSitemap()
             $can = $this->owner->getGooglePriority();
         }
 
-        $this->owner->invokeWithExtensions('alterCanIncludeInGoogleSitemap', $can);
+        // Allow override. In this case, since we don't know if this should be an "and" vs. "or" override, we'll use the LAST result (if any).
+        if ($override = $this->owner->invokeWithExtensions('alterCanIncludeInGoogleSitemap', $can)) $can = array_pop($override);
```
Comment (dhensby) [inline code/extensions/GoogleSitemapExtension.php]: the standard way to handle this in SS is to take a "min" of the results and return that. taking the last one is completely arbitrary

Comment (patricknelson) [inline code/extensions/GoogleSitemapExtension.php]: Min of what results? They're all booleans. And the concept (which I pointed out) was to take the last one to execute because there's no way to determine *which* should be used. How could we do what you're suggesting?

Comment (patricknelson) [inline code/extensions/GoogleSitemapExtension.php]: Note that if you translate that concept of minimum to booleans, that's a default false ("and") which is part of my point above.

Comment (dhensby) [inline code/extensions/GoogleSitemapExtension.php]: The idea being if ONE plugin says to deny the permission then it should be denied. A `min` is just a hacky way of finding a falsey val.

I guess really it should be a loop over the results as `null` returns should be ignored.

Comment (patricknelson) [inline code/extensions/GoogleSitemapExtension.php]: No need to ignore nulls, I believe those are never pushed into the array returned by `->extend()` (called by `->invokeWithExtensions()`), so we're good there. I think my problem was that I have a single site spread out across multiple domains and this plugin thinks it's being helpful by automatically assuming that it shouldn't be included. If there are multiple extensions tinkering with results, we have ourselves a little issue (then) -- of course at fault of the developer. In this case the question is how to handle that at a lower extension level.

Comment (patricknelson) [inline code/extensions/GoogleSitemapExtension.php]: p.s. @dhensby if you'd like, I will happily drop in that `min()` you're referring to, replacing `array_pop()` since it won't affect me personally. It's more of a theoretical debate at this point :smile:

**Comment by patricknelson** [comment]
@dhensby To help speed things along, I went ahead and implemented the `min()` change to supersede `array_pop()`. That way `false` will always take precedence. Hope this works! Thanks.

**Comment by dhensby** [comment]
Thanks @patricknelson - I'm happy with this - however your comment is now wrong :( Please can you update it?

I'd prefer if the condition used curly braces and went onto multiple lines, I think PSR-2 mandates that.

@wilr - any final thoughts?

**Comment by patricknelson** [comment]
Whoops @dhensby, nice catch! My apologies (I was in a hurry to wrap up for the day). I went ahead and clearly explained what's happening there and why!

**Comment by tractorcow** [comment]
> I'd prefer if the condition used curly braces and went onto multiple lines, I think PSR-2 mandates that.

+1

**Comment by tractorcow** [comment]
Also see https://github.com/silverstripe/silverstripe-framework/issues/3745 for the final invokeWithExtensions API for 4.x

**Comment by patricknelson** [comment]
Actually I think the bigger sin was the initialization of a variable within an if statement (something I typically tell myself I should not be doing). This avoids issues like what I've discovered in other SilverStripe modules (like this one https://github.com/bummzack/sortablefile/issues/26): 

```php
if ($override = $this->owner->invokeWithExtensions('alterCanIncludeInGoogleSitemap', $can)) $can = min($override);
```

Plus that line is so long, it makes sense to break it up into a second line as well (with curly braces) just for readability. 

I've gone ahead and updated the `if` statement to be all one line :smile: but here comes my debate/opinion on that subject (so the rest is all rant/fluff). 

----

Anyway, if you were to do this:

```php
$override = $this->owner->invokeWithExtensions('alterCanIncludeInGoogleSitemap', $can);
if ($override) $can = min($override);
```

It's much more readable to me (and here comes the opinion/debate), which I think is superior to the following, because it's so small and I don't think the curly syntax leave any ambiguity on what is happening here. This to me is needlessly verbose (and this is what I've submitted now in **this** PR):

```php
$override = $this->owner->invokeWithExtensions('alterCanIncludeInGoogleSitemap', $can);
if ($override) {
    $can = min($override);
}
```

Also note that the PSR-2 standard here only shows what an `if` statement should look like **with** `if`, `elseif` and `else` not with the `if` on its own. While it does make the general statement that "the body of each structure MUST be enclosed by braces," I would disagree under limited circumstances (**just** this one) based purely on aesthetics and readability, so long as the conditional statement is short. Albeit I will admit that at least as a standard, it's a good idea, only because each developer's opinion may vary depending on what constitutes A.) a long conditional statement and B.) a long block of code (e.g. my `$can = min($override)` above).

In conclusion... whilst methinks I doth protesteth too much, I will say that `MUST` should be downgraded to a `SHOULD` there in the opening paragraph of [Control Structures](http://www.php-fig.org/psr/psr-2/#5-control-structures) :-1:

**Comment by dhensby** [comment]
PSRs are actually just "recommendations" so it only matters if we want it to matter.

Thanks for amending this - to be honest I thought I'd already pulled it because I didn't think it was worth the fuss ;)

**Comment by patricknelson** [comment]
Hah no problem @dhensby I just had to vent.

**Comment by patricknelson** [comment]
@wilr @dhensby  I'm not seeing these updates yet in packagist. When I performed a reinstall via `composer require silverstripe/googlesitemaps` it installed `1.4.0`. Should this be tagged or are you not ready yet? Just curious. Thanks!

**Comment by wilr** [comment]
@patricknelson 1.5.0 tagged.

**Comment by patricknelson** [comment]
Cool, thanks!
